src/Propositions/Truthtable.hs

```diff
@@ -1,25 +1,29 @@
-﻿{- |
+﻿{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+{- |
 Description: Generation of truth tables for 'Logic's.
 License: GNU LGPLv3
 Maintainer: paul.bittner@uni-ulm.de
 
 Generation of truth tables for 'Logic's.
 -}
-module Truthtable where
+module Propositions.Truthtable where
 
-import Logic
+import Propositions.Logic
+import Data.Void
 
 -- | For a unary operator and a set of values, produces all results when applying that operator to that set of values.
-unarytable :: Logic a => (a -> a) -> [a] -> [a]
-unarytable op values = leval id . op <$> values
+unarytable :: (Logic a, Variable a ~ Void) => (a -> a) -> [a] -> [Value a]
+unarytable op values = leval absurd . op <$> values
 
 -- | Builds the cartesions product of two lists.
 cartesian :: [a] -> [b] -> [(a, b)]
 cartesian x y = (,) <$> x <*> y
 
 -- | For a binary operator and a set of values, produces all results when applying that operator to all combinations of values.
-binarytable :: Logic a => (a -> a -> a) -> [a] -> [a]
-binarytable op values = leval id . uncurry op <$> cartesian values values
+binarytable :: (Logic a, Variable a ~ Void) => (a -> a -> a) -> [a] -> [Value a]
+binarytable op values = leval absurd . uncurry op <$> cartesian values values
```

Comment:
The type system ensures, that `absurd` is never actually called with a valid value. This is achieved by the definition of `Void` as
```
data Void
```
> Note the absence of any type constructors.

There is no way to call `absurd`, because what would you pass it?
The only inhabitant of a data type such as `Void` is ⊥.

Applied to this usage:
The only problematic call to `binarytable` is something like `binarytable (\a b -> Variable bottom) lvalues` where `bottom` is `undefined` (in this case the program will be terminated with an error message) or an infinite loop like
```
bottom = bottom
```

Reply:
Ok, thanks for the explanation.